### PR TITLE
Fix CancelAsync Cause Deadlock

### LINKED ISSUE #1023: Is it possible to cancel SshCommand?
There is CancelAsync method but it seems to only cancel connection between client and server. Is there a way to stop the command from running? Similarly to CTRL+C while running in BASH shell.

### LINKED ISSUE #1147: CancelAsync does not cancel the command started by BeginExecute and blocks execution
When calling `CancelAsync` for a process started by `BeginExecute`, the process is not canceled and the result of the command cannot be retrieved.

With enough time for executing the command and without `CancelAsync` everything works fine:
```
            using (SshClient client = new SshClient(host, username, password))
            {
                client.Connect();
                using (SshCommand command = client.CreateCommand("sleep 10; echo run"))
                {
                    IAsyncResult task = command.BeginExecute();
                    Thread.Sleep(15000);
                    //command.CancelAsync();
                    var result = command.EndExecute(task);
                    System.Console.WriteLine(result);
                }
                client.Disconnect();
            }
```
After 15s, the following correctly appears in the console and the execution finishes:
```
run
```

If I reduce the sleep to 5s, and uncomment the `CancelAsync`:
```
            using (SshClient client = new SshClient(host, username, password))
            {
                client.Connect();
                using (SshCommand command = client.CreateCommand("sleep 10; echo run"))
                {
                    IAsyncResult task = command.BeginExecute();
                    Thread.Sleep(5000);
                    command.CancelAsync();
                    var result = command.EndExecute(task);
                    System.Console.WriteLine(result);
                }
                client.Disconnect();
            }
```
The process gets blocked at `EndExecute` instruction, it doesn't finish even though the command hasn't been canceled and it finishes after 10s from the beginning.

### PR DESCRIPTION
**Pull request by zeotuan**
Fix CancelAsync causing deadlocking:

it now signal WaitHandle of Cancelled event after sending SIGTERM/SIGKILL signal to try to abort the request.

close https://github.com/sshnet/SSH.NET/issues/1147 and close https://github.com/sshnet/SSH.NET/issues/1023

### CODE REVIEW

File: test/Renci.SshNet.IntegrationTests/OldIntegrationTests/SshCommandTest.cs
Comment (Rob-Hague) [inline test/Renci.SshNet.IntegrationTests/OldIntegrationTests/SshCommandTest.cs]: It cannot currently throw `OperationCanceledException` but perhaps it should?

```suggestion
            var asyncResult = cmd.BeginExecute();
            cmd.CancelAsync();
            Assert.ThrowsException<OperationCanceledException>(() => cmd.EndExecute(asyncResult));
```

File: test/Renci.SshNet.IntegrationTests/OldIntegrationTests/SshCommandTest.cs
Comment (Rob-Hague) [inline test/Renci.SshNet.IntegrationTests/OldIntegrationTests/SshCommandTest.cs]:
```suggestion
            Assert.AreEqual("", cmd.Result);
```

? (if not, the arguments should be reversed at least)

File: src/Renci.SshNet/SshCommand.cs
Comment (Rob-Hague) [inline src/Renci.SshNet/SshCommand.cs]: I think we should set `_asyncResult.IsCompleted = true` here as well (before this line)

File: src/Renci.SshNet/SshCommand.cs
Comment (Rob-Hague) [inline src/Renci.SshNet/SshCommand.cs:368]: Perhaps we drop the `Async` suffix here (and obsolete `CancelAsync`) as it is not an async method. @WojciechNagorski?

**Review by Rob-Hague — COMMENTED** [review]
Would it be possible to add a test for forceKill = true as well?

File: src/Renci.SshNet/SshCommand.cs
Comment (zeotuan) [reply src/Renci.SshNet/SshCommand.cs]: Sure. I am doing some further debugging. It seem that we might have to also execute the ```_callback``` as well as Flushing the ```OutputStream``` + ```ExntednedOutputStream``` .

On my local testing, ```Channel.Close``` event is never called and when I try to make ```SendExitSignalRequest``` wait for a response, I got ```Session Operation has Timeout``` Exception. Wondering if anyone else can recreate it.

Comment (Rob-Hague) [reply src/Renci.SshNet/SshCommand.cs]: Is it because of the call to `UnsubscribeFromEventsAndDisposeChannel(_channel);` in `EndExecute`?

Perhaps `_commandCancelledWaitHandle` should be removed? If I read https://datatracker.ietf.org/doc/html/rfc4254#section-6.10 correctly, we should always get SSH_MSG_CHANNEL_CLOSE after the signal, so it "should" just work.

edit: actually, not sure I did read it correctly

Comment (zeotuan) [reply src/Renci.SshNet/SshCommand.cs]: You are right! A proper implementation is to just send SIGTERM/KILL and wait for a response.

However, when I tried removing the _commandCancelledWaitHandle and made SendExitSignalRequest wait for a response, it never happened. If I just make SendExitSignalRequest send the message without waiting for a response, it will be blocked until the command actually finishes. 

My server openSSH version: OpenSSH_8.2p1 Ubuntu-4ubuntu0.11 so those signals are probably already supported.
I will takes a deeper look at the log.

File: src/Renci.SshNet/SshCommand.cs
Comment (Rob-Hague) [inline src/Renci.SshNet/SshCommand.cs]: I think the version at 43a57e12 is better. Spawning a thread to block on a wait is not ideal.

I am thinking we should not try and wait for a response after all (I'm not convinced there is one now).

As for the callback, I am not sure but I think if we have cancelled the operation, we should not execute the callback. (if one cancels a Task-based operation with a CancellationToken, then we get a TaskCanceledException and continuations are not executed)

File: src/Renci.SshNet/Common/SshOperationCancelledException.cs
Comment (Rob-Hague) [inline src/Renci.SshNet/Common/SshOperationCancelledException.cs:1]: Can we just use [System.OperationCanceledException](https://learn.microsoft.com/en-us/dotnet/api/system.operationcanceledexception)?

**Review by Rob-Hague — APPROVED** [review]

**Review by WojciechNagorski — APPROVED** [review]
Unfortunately, I don't have much time for this project at the moment. Only PRs that have @Rob-Hague's approval are verified and merged by me.

**Pull request merged**
